src/itx.rs

```diff
@@ -137,3 +139,277 @@ pub unsafe extern "C" fn inv_txfm_add_rust<BD: BitDepth>(
+            decl_itx_fn!([<dav1d_inv_txfm_add_identity_flipadst $wxh>], $bpc, $opt);
+        }
+    };
+}
+
+#[cfg(feature = "asm")]
+macro_rules! decl_itx17_fns {
+    ($wxh:ident, $opt:ident) => {
+        paste::paste! {
+            decl_itx16_fns!($wxh, $opt);
+            decl_itx_fn!([<dav1d_inv_txfm_add_wht_wht $wxh>], $opt);
+        }
+    };
+}
+
+#[cfg(feature = "asm")]
+macro_rules! decl_itx_fns {
+    ($ext:ident) => {
+        decl_itx17_fns!(_4x4, $ext);
+        decl_itx16_fns!(_4x8, $ext);
+        decl_itx16_fns!(_4x16, $ext);
+        decl_itx16_fns!(_8x4, $ext);
+        decl_itx16_fns!(_8x8, $ext);
+        decl_itx16_fns!(_8x16, $ext);
+        decl_itx2_fns!(_8x32, $ext);
+        decl_itx16_fns!(_16x4, $ext);
+        decl_itx16_fns!(_16x8, $ext);
+        decl_itx12_fns!(_16x16, $ext);
+        decl_itx2_fns!(_16x32, $ext);
+        decl_itx2_fns!(_32x8, $ext);
+        decl_itx2_fns!(_32x16, $ext);
+        decl_itx2_fns!(_32x32, $ext);
+        decl_itx_fn!(dav1d_inv_txfm_add_dct_dct_16x64, $ext);
+        decl_itx_fn!(dav1d_inv_txfm_add_dct_dct_32x64, $ext);
+        decl_itx_fn!(dav1d_inv_txfm_add_dct_dct_64x16, $ext);
+        decl_itx_fn!(dav1d_inv_txfm_add_dct_dct_64x32, $ext);
+        decl_itx_fn!(dav1d_inv_txfm_add_dct_dct_64x64, $ext);
+    };
+
+    ($bpc:ident, $ext:ident) => {
+        decl_itx16_fns!(_4x4, $bpc, $ext);
+        decl_itx16_fns!(_4x8, $bpc, $ext);
+        decl_itx16_fns!(_4x16, $bpc, $ext);
+        decl_itx16_fns!(_8x4, $bpc, $ext);
+        decl_itx16_fns!(_8x8, $bpc, $ext);
+        decl_itx16_fns!(_8x16, $bpc, $ext);
+        decl_itx2_fns!(_8x32, $bpc, $ext);
+        decl_itx16_fns!(_16x4, $bpc, $ext);
+        decl_itx16_fns!(_16x8, $bpc, $ext);
+        decl_itx12_fns!(_16x16, $bpc, $ext);
+        decl_itx2_fns!(_16x32, $bpc, $ext);
+        decl_itx2_fns!(_32x8, $bpc, $ext);
+        decl_itx2_fns!(_32x16, $bpc, $ext);
+        decl_itx2_fns!(_32x32, $bpc, $ext);
+        decl_itx_fn!(dav1d_inv_txfm_add_dct_dct_16x64, $bpc, $ext);
+        decl_itx_fn!(dav1d_inv_txfm_add_dct_dct_32x64, $bpc, $ext);
+        decl_itx_fn!(dav1d_inv_txfm_add_dct_dct_64x16, $bpc, $ext);
+        decl_itx_fn!(dav1d_inv_txfm_add_dct_dct_64x32, $bpc, $ext);
+        decl_itx_fn!(dav1d_inv_txfm_add_dct_dct_64x64, $bpc, $ext);
+    };
+}
+
+#[cfg(all(feature = "asm", any(target_arch = "x86", target_arch = "x86_64")))]
+extern "C" {
+    decl_itx_fns!(_avx512icl);
+    decl_itx_fns!(_10bpc, _avx512icl);
+    decl_itx_fns!(_avx2);
+    decl_itx_fns!(_10bpc, _avx2);
+    decl_itx_fns!(_12bpc, _avx2);
+    decl_itx_fns!(_sse4);
+    decl_itx_fns!(_ssse3);
+    decl_itx_fn!(dav1d_inv_txfm_add_wht_wht_4x4, _sse2);
+}
+
+#[cfg(all(feature = "asm", any(target_arch = "arm", target_arch = "aarch64")))]
+extern "C" {
+    decl_itx_fns!(_neon);
+}
+
+macro_rules! inv_txfm_fn {
+    ($type1:ident, $type2:ident, $w:literal, $h:literal, $shift:literal, $has_dconly:literal) => {
```

Comment:
I think `$has_dconly` can just be `expr`, not `literal`, since it's not used to make an `ident`.  And also, can we use `true` and `false` instead of `1` and `0`?  We can cast to `c_int` for now if we need to and fix that later, but for the literals, `true`/`false` is definitely much better.

Reply:
It could also be a `const` generic parameter instead I think.  That would be even better.  We can still generate separate `fn` ptrs out of it the same way we do with `::<BitDepth8>` now.

Reply:
Also, I think the same suggestions above apply also to `$shift`.

Reply:
Actually, since this is for the `_c_erased` ones, a bunch more can be made into `const` generic params instead.

Reply:
What's the advantage of making `shift` and `has_dconly` const generics? And are you talking about making it a generic parameter to `inv_txfm_add_rust` or on the generated `_c_erased` fn? I don't really see the advantage either way.

> I think `$has_dconly` can just be `expr`, not `literal`, since it's not used to make an `ident`.

Does that matter? Having it be a literal works just the same in this case.

Reply:
I've updated the macros to use `bool` literals instead of `1`/`0`.

Reply:
> What's the advantage of making `shift` and `has_dconly` const generics? And are you talking about making it a generic parameter to `inv_txfm_add_rust` or on the generated `_c_erased` fn? I don't really see the advantage either way.

I just thought it's slightly cleaner.  I looked at it more closely when doing #420 and I think it's overkill.

> > I think `$has_dconly` can just be `expr`, not `literal`, since it's not used to make an `ident`.
> 
> Does that matter? Having it be a literal works just the same in this case.

No, not really.